File: clients/ruby/lib/transport.rb
# frozen_string_literal: true

# Corndogs sync client transport: CSIL-RPC over a persistent TCP connection.
#
# This is Corndogs' own, ready-to-use carrier -- you do not need to write one or
# read any generator docs. Point it at a corndogs server's TCP address and pass
# it to the generated client:
#
#   require "corndogs"
#   require "corndogs/transport"
#
#   tr = TcpTransport.new("localhost:5080")
#   tr.start_heartbeat                      # keep the connection alive (background thread)
#   client = CorndogsClient.new(tr)
#   resp = client.submit_task(SubmitTaskRequest.new(
#     queue: "q", current_state: "submitted", auto_target_state: "sending",
#     timeout: -1, payload: "...".b, priority: 0,
#   ))
#
# The wire is CSIL-RPC framed with the canonical 4-byte big-endian length prefix
# over TCP. This carrier owns only the envelope + framing + connection; the
# generated client owns (de)serialization. Dependency-free (reuses the
# package's own CBOR codec).

require "socket"
require_relative "codec"

# Raised on a transport-level failure: a dropped connection, a non-zero
# transport status, or a malformed frame/envelope.
class TransportError < StandardError; end

# Raised when the server replies with a typed ServiceError (a status-0
# "ServiceError" variant). Carries the decoded #code and #message.
class CorndogsServiceError < StandardError
  attr_reader :code, :message

  def initialize(code, message)
    @code = code
    @message = message
    super("service error #{code}: #{message}")
  end
end

TAG_ENCODED_CBOR = 24 # RFC 8949 3.4.5.1 -- embedded encoded CBOR data item
CONTROL_SERVICE = "CorndogsService" # control-plane service name for the heartbeat
OP_PING = "$ping" # control-plane heartbeat op (never collides with app ops)
MAX_FRAME = 1025 << 20 # payload hard maximum plus RPC envelope allowance

# Sync CSIL-RPC transport over one persistent TCP connection. Calls are
# serialized (one in flight); the connection re-dials on failure. Thread-safe.
# Implements the generated client's transport seam: a duck-typed object
# responding to `call(service, op, req_bytes) -> resp_bytes`.
class TcpTransport
  # A ready TcpTransport for `addr` (host:port).
  def self.connect(addr)
    new(addr)
  end

  def initialize(addr, connect_timeout: 5)
    @host, @port = self.class.split_addr(addr)
    @connect_timeout = connect_timeout
    @mutex = Mutex.new
    @socket = nil
    @next_id = 0
    @hb_thread = nil
    @hb_stop = nil
  end

  def self.split_addr(addr)
    host, _, port = addr.to_s.rpartition(":")
    raise ArgumentError, "corndogs: address must be host:port, got #{addr.inspect}" if host.empty?

    [host, Integer(port)]
  end

  # The seam the generated client calls with the already-encoded request bytes.
  def call(service, op, req_bytes)
    @mutex.synchronize do
      sock = ensure_socket
      @next_id += 1
      env = CsilCbor.encode(
        {
          "v" => 1,
          "service" => service,
          "op" => op,
          "id" => @next_id,
          "payload" => CsilCbor::Tag.new(TAG_ENCODED_CBOR, req_bytes.b)
        }
      )
      begin
        write_frame(sock, env)
        resp = read_frame(sock)
      rescue TransportError
        reset
        raise
      rescue StandardError => e
        reset
        raise TransportError, e.message
      end
      if resp.nil?
        reset
        raise TransportError, "connection closed"
      end
      parse_response(resp)
    end
  end

  # --- heartbeat: one-shot, sync/blocking, and background thread -------------

  # Send one control-plane heartbeat (raises TransportError on a dead server).
  def ping
    call(CONTROL_SERVICE, OP_PING, "".b)
  end

  # SYNCHRONOUS heartbeat start: blocks, pinging every `interval` seconds,
  # until `stop` (a HeartbeatStop) is signalled -- or forever if no `stop` is
  # given. Run it in a Thread you own, or use `start_heartbeat` for the
  # background form.
  def run_heartbeat(interval = 15, stop: nil)
    loop do
      if stop
        return if stop.wait(interval)
      else
        sleep interval
      end
      ping
    end
  end

  # ASYNCHRONOUS heartbeat start: runs the heartbeat on a background Thread and
  # returns a stop proc. Idempotent -- only one background heartbeat runs at a
  # time; calling it again returns the existing stopper.
  def start_heartbeat(interval = 15)
    new_stop = nil
    @mutex.synchronize do
      return @hb_stop.stopper if @hb_thread

      new_stop = HeartbeatStop.new
      @hb_stop = new_stop
      @hb_thread = Thread.new { run_heartbeat(interval, stop: new_stop) }
      @hb_thread.abort_on_exception = false
    end
    new_stop.stopper
  end

  # Stops any heartbeat and closes the connection.
  def close
    hb_thread, hb_stop = nil, nil
    @mutex.synchronize do
      hb_thread, hb_stop = @hb_thread, @hb_stop
      @hb_thread = nil
      @hb_stop = nil
      reset
    end
    hb_stop&.set
    hb_thread&.join
  end

  private

  def ensure_socket
    return @socket if @socket

    sock = Socket.tcp(@host, @port, connect_timeout: @connect_timeout)
    sock.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
    @socket = sock
  end

  def reset
    @socket&.close
  rescue IOError
    nil
  ensure
    @socket = nil
  end

  def write_frame(sock, bytes)
    raise TransportError, "frame too large (#{bytes.bytesize})" if bytes.bytesize > MAX_FRAME

    sock.write([bytes.bytesize].pack("N") + bytes)
  end

  def read_frame(sock)
    head = read_exact(sock, 4)
    return nil unless head

    length = head.unpack1("N")
    raise TransportError, "frame too large (#{length})" if length > MAX_FRAME

    read_exact(sock, length)
  end

  def read_exact(sock, n)
    return "".b if n.zero?

    buf = "".b
    while buf.bytesize < n
      chunk = sock.read(n - buf.bytesize)
      return nil if chunk.nil?

      buf << chunk
    end
    buf
  end

  def parse_response(frame)
    val = CsilCbor.decode(frame)
    raise TransportError, "malformed response envelope" unless val.is_a?(Hash)

    status = val["status"] || 0
    raise TransportError, "transport status #{status}: #{val["error"]}" unless status.zero?

    payload = val["payload"]
    return "".b if payload.nil? # control replies (e.g. $pong) carry no payload

    payload = payload.value if payload.is_a?(CsilCbor::Tag)

    if val["variant"] == "ServiceError"
      inner = ServiceError.from_cbor(payload)
      raise CorndogsServiceError.new(inner.code, inner.message)
    end

    payload
  end
end

# A tiny stop-flag usable from `run_heartbeat`'s `stop:` kwarg or
# `start_heartbeat`'s background thread. #wait blocks up to `timeout` seconds
# and returns true if signalled meanwhile (or already signalled).
class HeartbeatStop
  def initialize
    @mutex = Mutex.new
    @cond = ConditionVariable.new
    @set = false
  end

  def set
    @mutex.synchronize do
      @set = true
      @cond.broadcast
    end
  end

  def wait(timeout)
    @mutex.synchronize do
      return true if @set

      @cond.wait(@mutex, timeout)
      @set
    end
  end

  def stopper
    method(:set)
  end
end

File: clients/ocaml/lib/transport.ml
(* Corndogs client transport: CSIL-RPC over a persistent TCP connection.
   ------------------------------------------------------------------------
   This is Corndogs' own, ready-to-use carrier — you do not need to write one
   or read csilgen docs. Point it at a corndogs server's TCP address and hand
   it to the generated client:

     let tr = Corndogs.Transport.connect "localhost:5080" in
     let _stop = Corndogs.Transport.start_heartbeat tr in  (* keep-alive, background *)
     let client = Corndogs.Client.make_client ~call:(Corndogs.Transport.call tr) in
     match Corndogs.Client.Corndogs_service.submit_task client req with
     | Ok resp -> ...
     | Error msg -> ...

   The wire is CSIL-RPC framed with the canonical 4-byte big-endian length
   prefix over TCP. This carrier owns only the envelope + framing + connection;
   the generated [Codec] owns (de)serialization. Dependency-free (stdlib
   [Unix] plus this package's own [Csil_cbor]).

   One connection, ONE call in flight at a time: [call] is guarded by a
   [Mutex], and the connection is dialed lazily and re-dialed on failure. This
   mirrors the official Python carrier's sync [TcpTransport] (OCaml's stdlib
   [Unix] sockets are blocking, so a multiplexed reader-thread design — as
   used by the Go carrier — is not the natural fit here). *)

module Cbor = Csil_cbor

let tag_encoded_cbor = 24 (* RFC 8949 §3.4.5.1 — embedded encoded CBOR data item *)
let max_frame = 1025 lsl 20 (* payload hard maximum plus RPC envelope allowance *)
let default_connect_timeout = 5.0 (* seconds *)
let default_heartbeat_interval = 15.0 (* seconds *)
let control_service = "CorndogsService"
let op_ping = "$ping" (* control-plane heartbeat op (never collides with app ops) *)

type t = {
  host : string;
  port : int;
  connect_timeout : float;
  mutex : Mutex.t; (* guards fd + next_id: one call in flight at a time *)
  mutable fd : Unix.file_descr option;
  mutable next_id : int64;
  mutable hb_stop : bool ref option; (* set while a background heartbeat thread runs *)
}

let split_addr (addr : string) : string * int =
  match String.rindex_opt addr ':' with
  | None -> failwith (Printf.sprintf "corndogs: address must be host:port, got %s" addr)
  | Some i ->
    let host = String.sub addr 0 i in
    let port_s = String.sub addr (i + 1) (String.length addr - i - 1) in
    (try (host, int_of_string port_s)
     with _ -> failwith (Printf.sprintf "corndogs: address must be host:port, got %s" addr))

(* [connect addr] builds a transport for "host:port". Dialing is lazy: no
   socket is opened until the first [call]. *)
let connect ?(connect_timeout = default_connect_timeout) (addr : string) : t =
  let host, port = split_addr addr in
  { host; port; connect_timeout; mutex = Mutex.create (); fd = None; next_id = 0L; hb_stop = None }

(* --- socket I/O -------------------------------------------------------- *)

let dial (host : string) (port : int) (timeout : float) : Unix.file_descr =
  let he =
    try Unix.gethostbyname host
    with Not_found -> failwith (Printf.sprintf "corndogs: cannot resolve host %s" host)
  in
  if Array.length he.Unix.h_addr_list = 0 then
    failwith (Printf.sprintf "corndogs: cannot resolve host %s" host);
  let sockaddr = Unix.ADDR_INET (he.Unix.h_addr_list.(0), port) in
  let fd = Unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
  (try
     Unix.set_nonblock fd;
     (try Unix.connect fd sockaddr with
      | Unix.Unix_error (Unix.EINPROGRESS, _, _) -> (
        let _, writable, _ = Unix.select [] [ fd ] [] timeout in
        match writable with
        | [] -> failwith "corndogs: connect timeout"
        | _ -> (
          match Unix.getsockopt_error fd with
          | Some err -> raise (Unix.Unix_error (err, "connect", host))
          | None -> ())));
     Unix.clear_nonblock fd;
     Unix.setsockopt fd Unix.TCP_NODELAY true;
     fd
   with e ->
     (try Unix.close fd with _ -> ());
     raise e)

let rec write_all (fd : Unix.file_descr) (b : bytes) (off : int) (len : int) : unit =
  if off < len then
    let n = try Unix.write fd b off (len - off) with Unix.Unix_error (Unix.EINTR, _, _) -> 0 in
    write_all fd b (off + n) len

(* Reads exactly [n] bytes, or returns [None] on a clean EOF before any byte of
   this read (a "connection closed" signal, mirroring the Go/Python carriers). *)
let read_exact (fd : Unix.file_descr) (n : int) : bytes option =
  let buf = Bytes.create n in
  let rec loop off =
    if off >= n then Some buf
    else
      let r = try Unix.read fd buf off (n - off) with Unix.Unix_error (Unix.EINTR, _, _) -> -1 in
      if r = 0 then None
      else if r < 0 then loop off
      else loop (off + r)
  in
  loop 0

let write_frame (fd : Unix.file_descr) (payload : bytes) : unit =
  let len = Bytes.length payload in
  if len > max_frame then failwith (Printf.sprintf "corndogs: frame too large (%d)" len);
  let prefix = Bytes.create 4 in
  Bytes.set prefix 0 (Char.chr ((len lsr 24) land 0xff));
  Bytes.set prefix 1 (Char.chr ((len lsr 16) land 0xff));
  Bytes.set prefix 2 (Char.chr ((len lsr 8) land 0xff));
  Bytes.set prefix 3 (Char.chr (len land 0xff));
  write_all fd prefix 0 4;
  write_all fd payload 0 len

let read_frame (fd : Unix.file_descr) : bytes option =
  match read_exact fd 4 with
  | None -> None
  | Some hdr ->
    let byte i = Char.code (Bytes.get hdr i) in
    let len = (byte 0 lsl 24) lor (byte 1 lsl 16) lor (byte 2 lsl 8) lor byte 3 in
    if len > max_frame then failwith (Printf.sprintf "corndogs: frame too large (%d)" len);
    read_exact fd len

(* --- envelope ------------------------------------------------------------ *)

let build_request ~(service : string) ~(op : string) ~(id : int64) ~(payload : bytes) : bytes =
  Cbor.encode
    (Cbor.Map
       [
         (Cbor.Text "v", Cbor.Uint 1L);
         (Cbor.Text "service", Cbor.Text service);
         (Cbor.Text "op", Cbor.Text op);
         (Cbor.Text "id", Cbor.Uint id);
         (Cbor.Text "payload", Cbor.Tag (tag_encoded_cbor, Cbor.Bytes payload));
       ])

let map_get (kvs : (Cbor.t * Cbor.t) list) (key : string) : Cbor.t option =
  List.find_map (fun (k, v) -> if k = Cbor.Text key then Some v else None) kvs

(* Decodes one response frame into the inner reply payload, or a
   human-readable error string. A non-zero transport [status] or a
   "ServiceError" [variant] both surface as [Error]. *)
let parse_response (frame : bytes) : (bytes, string) result =
  match Cbor.decode frame with
  | Error e -> Error ("corndogs: decode response envelope: " ^ e)
  | Ok (Cbor.Map kvs) -> (
    let status = match map_get kvs "status" with Some v -> ( try Cbor.to_i64 v with _ -> 0L) | None -> 0L in
    if Int64.compare status 0L <> 0 then
      let msg = match map_get kvs "error" with Some v -> ( try Cbor.to_text v with _ -> "") | None -> "" in
      Error (Printf.sprintf "corndogs: transport status %Ld: %s" status msg)
    else
      match map_get kvs "payload" with
      | None -> Ok Bytes.empty (* control replies (e.g. $pong) carry no payload *)
      | Some pv -> (
        let pv = match pv with Cbor.Tag (_, inner) -> inner | other -> other in
        match pv with
        | Cbor.Bytes inner -> (
          match map_get kvs "variant" with
          | Some (Cbor.Text "ServiceError") -> (
            try
              let se = Codec.decode_service_error_bytes inner in
              Error (Printf.sprintf "corndogs: service error %Ld: %s" se.Types.code se.Types.message)
            with Failure m -> Error ("corndogs: undecodable ServiceError payload: " ^ m))
          | _ -> Ok inner)
        | _ -> Error "corndogs: response payload is not a tag-24 byte string"))
  | Ok _ -> Error "corndogs: malformed response envelope"

(* --- connection lifecycle ------------------------------------------------ *)

let ensure_fd (t : t) : Unix.file_descr =
  match t.fd with
  | Some fd -> fd
  | None ->
    let fd = dial t.host t.port t.connect_timeout in
    t.fd <- Some fd;
    fd

let reset (t : t) : unit =
  (match t.fd with
   | Some fd -> ( try Unix.close fd with _ -> ())
   | None -> ());
  t.fd <- None

(* [call t ~service ~op ~payload] performs one CSIL-RPC round trip and returns
   the inner reply payload bytes (or an error). Matches the generated
   [Client.client]'s [call] seam exactly, so `Transport.call tr` partially
   applies straight into [Client.make_client ~call:(Transport.call tr)].
   Thread-safe: calls are serialized (one in flight); a failed call tears
   down the connection so the next call re-dials. *)
let call (t : t) ~(service : string) ~(op : string) ~(payload : bytes) : (bytes, string) result =
  Mutex.lock t.mutex;
  Fun.protect
    ~finally:(fun () -> Mutex.unlock t.mutex)
    (fun () ->
      match
        (try Ok (ensure_fd t) with
         | Failure m -> Error m
         | Unix.Unix_error (e, fn, _) -> Error (Printf.sprintf "corndogs: %s: %s" fn (Unix.error_message e)))
      with
      | Error e -> Error e
      | Ok fd -> (
        t.next_id <- Int64.add t.next_id 1L;
        let id = t.next_id in
        let env = build_request ~service ~op ~id ~payload in
        try
          write_frame fd env;
          match read_frame fd with
          | None ->
            reset t;
            Error "corndogs: connection closed"
          | Some resp -> parse_response resp
        with
        | Failure m ->
          reset t;
          Error m
        | Unix.Unix_error (e, fn, _) ->
          reset t;
          Error (Printf.sprintf "corndogs: %s: %s" fn (Unix.error_message e))))

(* --- heartbeat: one-shot, sync (blocking), and async (background thread) - *)

(* Sends a single control-plane heartbeat; [Error _] means the server is
   unreachable. Cheap; keeps an idle connection alive and detects a dead
   server. *)
let ping (t : t) : (unit, string) result =
  match call t ~service:control_service ~op:op_ping ~payload:Bytes.empty with
  | Ok _ -> Ok ()
  | Error e -> Error e

(* SYNCHRONOUS heartbeat start: blocks, pinging every [interval] seconds,
   until [stop] is set to [true] (returns [None]) or a ping fails (returns
   [Some error]). Run it in a [Thread] you own, or use [start_heartbeat] for
   the background form. *)
let run_heartbeat ?(interval = default_heartbeat_interval) ?(stop : bool ref option) (t : t) : string option =
  let stopped () = match stop with Some r -> !r | None -> false in
  let rec loop () =
    if stopped () then None
    else begin
      Thread.delay interval;
      if stopped () then None
      else
        match ping t with
        | Ok () -> loop ()
        | Error e -> Some e
    end
  in
  loop ()

(* ASYNCHRONOUS heartbeat start: runs the heartbeat on a background [Thread]
   and returns a stop function. Only one background heartbeat runs at a
   time — calling this again while one is running returns a stopper for the
   existing one. *)
let start_heartbeat ?(interval = default_heartbeat_interval) (t : t) : unit -> unit =
  match t.hb_stop with
  | Some existing -> fun () -> existing := true
  | None ->
    let stop = ref false in
    t.hb_stop <- Some stop;
    ignore (Thread.create (fun () -> ignore (run_heartbeat ~interval ~stop t)) ());
    fun () ->
      stop := true;
      t.hb_stop <- None

(* Shuts the transport down: stops any background heartbeat and closes the
   connection. *)
let close (t : t) : unit =
  (match t.hb_stop with
   | Some stop ->
     stop := true;
     t.hb_stop <- None
   | None -> ());
  Mutex.lock t.mutex;
  reset t;
  Mutex.unlock t.mutex

(* Convenience: wires a ready [Client.client] over a fresh transport for
   [addr] (host:port). Equivalent to [Client.make_client ~call:(call
   (connect addr))], returned alongside the transport so callers can still
   drive the heartbeat / [close] it. *)
let client (t : t) : Client.client = Client.make_client ~call:(call t)
